feat(`storage`): add support for eip 7702 bytecode

EdwardJES opened the pull request:
PR addressing https://github.com/risechain/pevm/issues/317, working towards https://github.com/risechain/pevm/issues/302

EdwardJES commented:
Dam, the new `bytecodes` file has balooned in size due to the new enum field bringing the size to ~ 53mb and GH is throwing warnings 😭 .

hai-rise commented on src/storage.rs:
```diff
@@ -68,26 +68,49 @@ impl Default for AccountBasic {
 
+#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
+pub struct Eip7702Code {
+    /// EOA which which will inherit the bytecode.
```
Double `which`!

hai-rise commented on src/storage.rs:
```diff
@@ -68,26 +68,49 @@ impl Default for AccountBasic {
 #[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
-pub struct EvmCode {
-    /// Bytecode with 32 zero bytes padding
+pub struct LegacyAnalyzedCode {
```
We can just call this `LegacyCode` that we enforce to be analyzed for performance.

hai-rise commented on src/storage.rs:
```diff
@@ -68,26 +68,49 @@ impl Default for AccountBasic {
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum EvmCode {
+    /// Analyzed legacy code, with jump table.
+    LegacyAnalyzed(LegacyAnalyzedCode),
```
Just `Legacy` is good as we should map both non & analyzed REVM bytecodes to the same analyzed structure.

hai-rise commented on src/storage.rs:
```diff
@@ -68,26 +68,49 @@ impl Default for AccountBasic {
-            Bytecode::new_analyzed(code.bytecode, code.original_len, JumpTable(code.jump_table))
+        match code {
+            EvmCode::Eip7702(code) => {
+                Bytecode::Eip7702(Eip7702Bytecode::new(code.delegated_address))
```
Shouldn't we construct a full `Bytecode::Eip7702` by "forwarding" all the fields here? This conversion seems to create an empty code regardless of the current `code`.

hai-rise commented on data/bytecodes.bincode:
Hmm, I wonder what changes here?

EdwardJES replied:
Good question. This is because the bincode encoding has to support the new field I added. I guess this is to be aware when decoding.

hai-rise replied:
Ah, yes, it should add a prefix byte to know which arm/type this bytecode is. Fair, thanks 🙏.

hai-rise reviewed (changes requested):
This is very neat, thanks! 

I've left a few comments and adding a test would help a lot 🙏.

hai-rise commented on src/storage.rs:
```diff
@@ -68,26 +68,49 @@ impl Default for AccountBasic {
-            Bytecode::new_analyzed(code.bytecode, code.original_len, JumpTable(code.jump_table))
+        match code {
+            EvmCode::Eip7702(code) => {
+                Bytecode::Eip7702(Eip7702Bytecode::new(code.delegated_address))
```
I think there are two choices here:
1. For simplicity, make `EvmCode::Eip7702` wrap an `Address` and forward it here. Then perhaps add a `TODO` to support custom versions and magics.
2. Construct the full `Bytecode::Eip7702` by forwarding all the fields here.

Which do you prefer?

EdwardJES replied:
Good callout. I'll go for the latter here.

hai-rise commented on src/storage.rs:
```diff
@@ -68,26 +68,49 @@ impl Default for AccountBasic {
+    /// Version of the bytecode.
+    version: u8,
+    /// Raw bytecode encoding MAGIC, version and delegated address.
+    raw: Bytes,
```
I don't think we need this for "storage". We can always re-construct it at runtime for execution (like at `revm` level) from `address` and `version`.

EdwardJES replied:
Good callout. I've removed `raw` and kept `version` so we can preserve it. This has naturally made me choose option 2. from above.

EdwardJES commented:
Hey @hai-rise, thanks for the feedback 🙏 . I've added tests and addressed the above comments. LMK if you'd like me to add tests converting back to the `revm` types.

EdwardJES commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,65 @@
+}
+
+#[test]
+fn test_evmcode_from_revm_bytecode_legacy_raw() {
```
I'm in favour of the format: `test_<describe the test>` and it would be great to consistency across the test names. Please lmk what you think.

hai-rise replied:
I don't have a strong preference here. Feel the `test_` prefix is a bit redundant as there's usually a `#[test]` above the `fn`, and the log would be:
```
running 3 tests
test test_evmcode_from_revm_bytecode_eip7702 ... ok
test test_evmcode_from_revm_bytecode_legacy_analyzed ... ok
test test_evmcode_from_revm_bytecode_legacy_raw ... ok
```

Instead of just:
```
running 2 tests
test beneficiary_heavy_evaluation ... ok
test beneficiary_random ... ok
```

EdwardJES replied:
No problems, will adjust.

EdwardJES commented on src/storage.rs:
```diff
@@ -68,16 +71,34 @@ impl Default for AccountBasic {
-    original_len: usize,
+pub struct LegacyCode {
+    /// Bytecode with 32 zero bytes padding.
+    pub bytecode: Bytes,
```
`pub(crate)`?

EdwardJES commented on src/storage.rs:
```diff
@@ -68,16 +71,34 @@ impl Default for AccountBasic {
-/// EVM Code, currently mapping to REVM's [ByteCode::LegacyAnalyzed].
-// TODO: Support raw legacy & EOF
+/// Analyzed legacy code.
 #[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
```
Is `PartialEq` needed here 🤔

hai-rise replied:
Yes, it is required for `EvmCode` -> `EvmAccount` -> `EvmStateTransitions` -> `PevmTxExecutionResult`. It won't be "required" anymore after #202 but comparing bytecodes still sounds convenient. I guess we can also implement more performant `PartialEq`s by comparing only the raw bytecodes, but it's out-of-scope here :pray:.

hai-rise replied:
We can remove the `Default`s though, which is also error-prone. Anything that can have an `EvmCode` like `EvmAccount` should use `None` instead of a default code (that is probably invalid anyway).

EdwardJES replied:
Thank you, makes sense.

hai-rise commented on src/storage.rs:
```diff
@@ -68,16 +71,34 @@ impl Default for AccountBasic {
-    /// Original bytes length
-    original_len: usize,
+pub struct LegacyCode {
+    /// Bytecode with 32 zero bytes padding.
```
The `Eip7702Code`'s raw bytecode conversation made me realize we don't need to store padded bytecodes as well. Perhaps add a `// TODO: Store unpadded bytecode and pad on revm conversion` here?

hai-rise commented on src/storage.rs:
```diff
@@ -68,16 +71,34 @@ impl Default for AccountBasic {
-    original_len: usize,
+pub struct LegacyCode {
+    /// Bytecode with 32 zero bytes padding.
+    pub bytecode: Bytes,
```
Let's not make these fields `pub` as that allows people to construct an invalid `LegacyCode` in memory. For now, let's remove the `Default`s and guard the only entry of `revm` conversion well. In the future, we can add `new` functions that do proper validation and analysis and return a `Result`.

EdwardJES replied:
Good callout.

hai-rise commented on src/storage.rs:
```diff
@@ -86,23 +107,39 @@ impl From<EvmCode> for Bytecode {
                 jump_table: code.jump_table.0,
-            },
+            }),
             Bytecode::Eof(_) => unimplemented!("TODO: Support EOF"),
```
Nit but putting `unimplemented!` last is more visible :pray:.

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+use std::str::FromStr;
```
I think this is cleaner as a 
```
#[cfg(test)]
mod tests {
    use super::*;
}
```

in `storage.rs`. It's more appropriate for unit tests for location, removing the need to make fields `pub`, etc. Most stuff in `tests` are integration tests for the `execute_*` functions.

EdwardJES replied:
Good call, not sure what I was thinking here. These are definitely unit tests I'm targetting.

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+    primitives::{Bytecode, Eip7702Bytecode},
+};
+
+const BYTECODE: alloy_primitives::Bytes = bytes!("608060405234801561001057600080fd5b506004361061002b5760003560e01c8063920a769114610030575b600080fd5b61004361003e366004610374565b610055565b60405190815260200160405180910390f35b600061006082610067565b5192915050565b60606101e0565b818153600101919050565b600082840393505b838110156100a25782810151828201511860001a1590930292600101610081565b9392505050565b825b60208 ...
```
A comment on what this bytecode is would be nice :pray:.

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+const BYTECODE: alloy_primitives::Bytes = bytes!("608060405234801561001057600080fd5b506004361061002b5760003560e01c8063920a769114610030575b600080fd5b61004361003e366004610374565b610055565b60405190815260200160405180910390f35b600061006082610067565b5192915050565b60606101e0565b818153600101919050565b600082840393505b838110156100a25782810151828201511860001a1590930292600101610081565b9392505050565b825b60208 ...
+
+#[test]
+fn test_evmcode_from_revm_bytecode_eip7702() {
```
Nit but ordering-wise I'd put this last out of 3 functions :sweat_smile:.

hai-rise replied:
An interesting test we can add is to mutate an `Eip7702Bytecode`'s `version`, convert it to `EvmCode` and assert that the `version` is correctly kept. Would raise an error if a future change hardcodes `version` or something.

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+    let evmcode = EvmCode::from(bytecode);
+    assert!(
+        matches!(evmcode, EvmCode::Eip7702(Eip7702Code { delegated_address, .. })
+            if delegated_address == addr
```
We can also assert that `version == EIP7702_VERSION` here?

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+    );
+
+    // New fromn raw.
+    let byte_str = format!("ef0100{}", addr.to_string().trim_start_matches("0x"));
```
We can use `EIP7702_MAGIC_BYTES` and `EIP7702_VERSION` instead of hard-coded values here?

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+    let evmcode = EvmCode::from(bytecode);
+    assert!(
+        matches!(evmcode, EvmCode::Eip7702(Eip7702Code { delegated_address, version })
+            if delegated_address == addr && version == 0
```
`EIP7702_VERSION`?

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+            )
+        );
+    } else {
+        panic!("Expected LegacyAnalyzed Bytecode")
```
We can write something like this to deduplicate code and make these tests much cleaner:
```rust
fn eq_bytecodes(revm_code: &Bytecode, pevm_code: &EvmCode) -> bool {
    match (revm_code, pevm_code) {
        ...
    }
}
```

hai-rise commented on tests/bytecode.rs:
```diff
@@ -0,0 +1,73 @@
+}
+
+#[test]
+fn test_evmcode_from_revm_bytecode_legacy_analyzed() {
```
We can probably merge this with `_raw` to just `_legacy()`.

EdwardJES commented on src/storage.rs:
```diff
@@ -68,41 +71,73 @@ impl Default for AccountBasic {
 impl From<EvmCode> for Bytecode {
     fn from(code: EvmCode) -> Self {
         // TODO: Better error handling.
-        // A common trap would be converting a default [EvmCode] into
```
Rm'd this comment as `default` is no longer derived.

hai-rise replied:
Nice!

EdwardJES commented:
@hai-rise, thank you for the feedback. Have refactored with the suggestion above 🙏

hai-rise reviewed (changes requested):
A few more nits :pray:.

EdwardJES commented on src/storage.rs:
```diff
@@ -193,3 +228,78 @@ mod in_memory;
+        }
+    }
+
+    // Bytecode from Storage.sol.
```
Yeah apologies, I was thinking that the `storage.sol` reference from the default remix contract would've done the trick. Agree, not clear. I can make this clearer.

hai-rise replied:
It is much better now, thanks!

EdwardJES commented on src/storage.rs:
```diff
@@ -193,3 +228,78 @@ mod in_memory;
+        }
+    }
+
+    fn eq_eip7702_version_addr(evmcode: &EvmCode, address: &Address, version: u8) -> bool {
```
Good point.

EdwardJES commented on src/storage.rs:
```diff
@@ -193,3 +228,78 @@ mod in_memory;
+
+        // Assert that From<Bytecode> preserves version.
+        let new_version = 1;
+        let mut eip_bytecode = Eip7702Bytecode::new_raw(bytes.into()).unwrap();
```
great catch

EdwardJES commented on src/storage.rs:
```diff
@@ -193,3 +228,78 @@ mod in_memory;
+    use super::*;
+
+    #[test]
+    fn evmcode_from_revm_bytecode_legacy() {
```
Good idea. 

I've decided to lean on the side of caution here and do this after each `eq_bytecodes`, even though they will all fall to the same match arm. Feels safer going with this approach for evolvability and rigour. 

The alternative is to only do it once in `legacy_bytecodes` and `eip7702_bytecodes`.

Obvious downside with my approach is the like of in-lining and less terse code. 

WDYT?

hai-rise replied:
This works for me too, it also helps scope each "sub-section" well :).

EdwardJES commented on src/storage.rs:
```diff
@@ -193,3 +228,78 @@ mod in_memory;
+
+        // New from address.
+        let bytecode = Bytecode::Eip7702(Eip7702Bytecode::new(addr));
+        let evmcode = EvmCode::from(bytecode.clone());
```
Under the assumption that `assert_equals!` is used after each `eq_bytecodes`, I've not inlined here, as I felt it looked bit unwieldy if I did. 

```
 assert!(eq_bytecodes(
            &Bytecode::Eip7702(Eip7702Bytecode::new(delegated_address)),
            &EvmCode::from(Bytecode::Eip7702(Eip7702Bytecode::new(delegated_address)))
        ));
        assert_eq!(
            Bytecode::Eip7702(Eip7702Bytecode::new(delegated_address)),
            Bytecode::from(EvmCode::from(Bytecode::Eip7702(Eip7702Bytecode::new(
                delegated_address
            ))))
        );
 ```
 
 As opposed to:
 
 ```
        let bytecode = Bytecode::Eip7702(Eip7702Bytecode::new(delegated_address));
        let evm_code = EvmCode::from(Bytecode::Eip7702(Eip7702Bytecode::new(delegated_address)));
        assert!(eq_bytecodes(&bytecode, &evm_code));
        assert_eq!(bytecode, Bytecode::from(evm_code));
```

EdwardJES replied:
Happy to revert to inlining if you prefer only doing `assert_eq!` once per call to `eq_bytecodes`

hai-rise replied:
Yes, it's better to reuse `evm_code` in the two asserts now!

hai-rise commented on src/storage.rs:
```diff
@@ -193,3 +227,74 @@ mod in_memory;
+    use super::*;
+
+    // Bytecode from Forge's default Counter.sol contract, compiled with solc 0.8.13.
+    // https://github.com/foundry-rs/foundry/blob/nightly-fe2acca4e379793539db80e032d76ffe0110298b/testdata/multi-version/Counter.sol
```
This is much better, thanks!

hai-rise reviewed (approved):
Beautiful!
